crates/hir_ty/src/diagnostics/expr.rs

```diff
@@ -150,20 +154,54 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
         }
     }
 
-    fn validate_call(&mut self, db: &dyn HirDatabase, call_id: ExprId, expr: &Expr) -> Option<()> {
+    fn check_for_filter_map_next(&mut self, db: &dyn HirDatabase) {
```

Comment:
This adds another iteration through the exprs in body. I could incorporate this logic into the loop in validate_body at the cost of making that function harder to follow.